File: templates/Provider.cpp
// #######################################################################
// NOTICE
//
// This software (or technical data) was produced for the U.S. Government
// under contract, and is subject to the Rights in Data-General Clause
// 52.227-14, Alt. IV (DEC 2007).
//
// Copyright 2019 The MITRE Corporation. All Rights Reserved.
// #######################################################################

#include <NewProvider.h>
#include <fstream>
#include <map>

NewProvider::NewProvider()
{
    // Initialize metadata
    std::string biqt_home = getenv("BIQT_HOME");
    std::ifstream desc_file(biqt_home +
                                "/providers/NewProvider/descriptor.json",
                            std::ifstream::binary);
    desc_file >> DescriptorObject;

    // TODO: Instantiate any additional objects needed for your algorithm
}

NewProvider::~NewProvider()
{
    // TODO: If needed, populate the destructor
}

Provider::EvaluationResult NewProvider::evaluate(const std::string &file)
{
    // Initialize some variables
    Provider::EvaluationResult evalResult;
    Provider::QualityResult qualityResult;

    // TODO: Read the input image 'file' (string indicating file path to image)
    
	// TODO: Populate metrics and features maps.	
	qualityResult.metrics["quality"] = 555.55;
	qualityResult.features["left_eye_x"] = 1;
	
	evalResult.qualityResult.push_back(std::move(qualityResult));
    return evalResult;
}

DLL_EXPORT const char *provider_eval(const char *cFilePath)
{
    NewProvider p;
    std::string filePath(cFilePath);
    Provider::EvaluationResult result = p.evaluate(filePath);
    return Provider::serializeResult(result);
}

File: java/src/main/java/org/mitre/biqt/Provider.java
// #######################################################################
// NOTICE
//
// This software (or technical data) was produced for the U.S. Government
// under contract, and is subject to the Rights in Data-General Clause
// 52.227-14, Alt. IV (DEC 2007).
//
// Copyright 2019 The MITRE Corporation. All Rights Reserved.
// #######################################################################

package org.mitre.biqt;

import java.util.List;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;

public abstract class Provider {
	private String providerName;
	private String message;
	private int errorCode;

	protected Provider() {
		this.providerName = this.getClass().getSimpleName();
	}

	/**
	 * Evaluates an image file and returns a JSON String
	 *
	 * @param filename The path to the file to be evaluated.
	 *
	 * @return A JSON String representation of the result
	 */
	public JSONObject evaluate(String filename) {
		JSONObject jsonResult = new JSONObject();
		JSONObject providerObject = new JSONObject();
		JSONObject resultObject = new JSONObject();
		JSONArray resultNameArray = new JSONArray();
		JSONObject features = new JSONObject();
		JSONObject metrics = new JSONObject();
		resultNameArray.add(resultObject);
		resultObject.put("features", features);
		resultObject.put("metrics", metrics);
		providerObject.put(this.providerName, resultObject);
		this.setResults(filename, features, metrics);
		jsonResult.put("errorCode", this.errorCode);
		jsonResult.put("message", this.message);
		jsonResult.put("provider", this.providerName);
		jsonResult.put("qualityResult", providerObject);

		return jsonResult;
	}

	protected abstract void setResults(String filename, JSONObject features,
									   JSONObject metrics);

	public static List<Class> getProviders() {
		return null;
	}
}
